### PR TITLE
PoR

### PR DESCRIPTION
**Pull request by bayram98**
# Description

This PR is ready to review. Main changes with this PR is to make a separate directory to `POR`, where it will be running upon request.  

How to run PoR

1. Define env variables with new addition `POR_AGGREGATOR_HASH`
2. Build: `yarn build`
3. Run por: `yarn start:por`

Changes:
- new env variable `POR_AGGREGATOR_HASH`
- new directory for PoR, includes fetching and reporting of data.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.

## Deployment

- [ ] Should publish npm package
- [ ] Should publish Docker image

### CODE REVIEW

File: core/src/por/api.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import axios from 'axios'
+import { OraklError, OraklErrorCode } from '../errors'
+import { CHAIN, ORAKL_NETWORK_API_URL } from '../settings'
+import { IAggregator } from '../types'
+import { buildUrl } from '../utils'
+import { IData } from './types'
+
+export async function loadAggregator({ aggregatorHash }: { aggregatorHash: string }) {
```
Comment (martinkersner) [inline core/src/por/api.ts:8]: @bayram98 If I understand it correctly, we are not going to activate the aggregator. Am I right?

File: core/src/por/api.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  const chain = CHAIN
+  try {
+    const url = buildUrl(ORAKL_NETWORK_API_URL, `aggregator/${aggregatorHash}/${chain}`)
```
Comment (martinkersner) [inline core/src/por/api.ts:11]: We could pass the `CHAIN` variable directly without creating a transitory variable.

```suggestion
    const url = buildUrl(ORAKL_NETWORK_API_URL, `aggregator/${aggregatorHash}/${CHAIN}`)
```

File: core/src/por/fetcher.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    const reducer = reducerMapping[r.function]
+    if (!reducer) {
+      throw new PorError(PorErrorCode.InvalidReducer)
+    }
+    return reducer(r?.args)
+  })
+}
+
+async function fetchData(feed) {
+  const rawDatum = await (await axios.get(feed.url)).data
```
Comment (martinkersner) [inline core/src/por/fetcher.ts:41]: We should make `try-catch` block here.

File: core/src/por/reducer.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
+import { PorError, PorErrorCode } from './errors'
+
+export const DATA_FEED_REDUCER_MAPPING = {
```
Comment (martinkersner) [inline core/src/por/reducer.ts:3]: Can't we just import the reducer, rather than copy it?

File: fetcher/src/job/job.errors.ts [hunk 5]
```diff
@@ -8,7 +8,6 @@ export class FetcherError extends Error {
 }
 
 export enum FetcherErrorCode {
-  IncompleteDataFeed,
```
Comment (martinkersner) [inline fetcher/src/job/job.errors.ts:11]: Is this update somehow related to the PoR?

File: core/src/por/types.ts [hunk 6]
```diff
@@ -0,0 +1,6 @@
+export interface IData {
```
Comment (martinkersner) [inline core/src/por/types.ts:1]: If possible, it would be also good if we can reuse the implementation from other files. We don't have to do it now, but we should at least make an issue for it.

File: core/src/por/reporter.ts [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+}) {
+  const contract = new ethers.Contract(aggregator.address, Aggregator__factory.abi, PROVIDER)
+  const latestRoundData = await contract.latestRoundData()
+
+  // Check Submission Hearbeat
+  const updatedAt = Number(latestRoundData.updatedAt) * 1000 // convert to milliseconds
+  const timestamp = Date.now()
+  const heartbeat = aggregator.heartbeat
+
+  if (updatedAt + heartbeat < timestamp) {
```
Comment (martinkersner) [inline core/src/por/reporter.ts:33]: Minor comment, it would make more sense to call it `now` rather than `timestamp`.

**Review by martinkersner — APPROVED** [review]
LGTM! I left comments, but they are more recommendation and we can fix it in next PR. It is usually related to reuse of implementation from other files so we do not copy & paste code.

BTW I notice, you store data only in `data` table. It would be better to store it also in `aggregate` table which is then used for visualizations.

File: core/src/por/api.ts (hunk 1 above)
Comment (bayram98) [reply core/src/por/api.ts:8]: Yes.

File: fetcher/src/job/job.errors.ts (hunk 5 above)
Comment (bayram98) [reply fetcher/src/job/job.errors.ts:11]: This is from fetcher. I realized that we did not used it anywhere tried to clean-up.

Comment (martinkersner) [reply fetcher/src/job/job.errors.ts:11]: Okay! Good good!

File: core/src/por/reducer.ts (hunk 4 above)
Comment (nick-bisonai) [reply core/src/por/reducer.ts:3]: would it be okay to import reducer.ts from `./src/worker/reducer.ts`? thought por can't refer from `worker` foler

File: core/src/por/reporter.ts [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+  // Check deviation threashold
+  if (aggregator.threshold && latestRoundData.answer) {
+    const latestSubmission = Number(latestRoundData.answer)
+    const currentSubmission = Number(value)
+
+    const range = latestSubmission * aggregator.threshold
+    const l = currentSubmission - range
+    const r = currentSubmission + range
```
Comment (nick-bisonai) [inline core/src/por/reporter.ts:46]: i'm not sure if i got it correctly, but shouldn't this be
```
    const l = latestSubmission - range
    const r = latestSubmission + range
```

Comment (bayram98) [reply core/src/por/reporter.ts:46]: @nick-bisonai Yes you are right. I will fix it now.

**Review by jay-bisonai — APPROVED** [review]
LGTM

File: fetcher/src/job/job.errors.ts (hunk 5 above)
Comment (nick-bisonai) [reply fetcher/src/job/job.errors.ts:11]: it would be appreciated if you could make a separate PR for refactoring from next time :)

**Comment by bayram98** [comment]
@martinkersner @nick-bisonai I will merge the PR now, and will make a new PR for the recommend changes.
